### PR TITLE
VAGOV-4311: Add custom menu perm module.

### PR DESCRIPTION
**Pull request by ethanteague**
## Description
See #4311

## Testing done
Visual / behat / phpunit

## Screenshots
image
image
image
image
image
image
image
image

## QA steps
- As an administrator, go to `admin/config/content/va_gov_menu_access`
- Check the following roles: `Content creator - VAMC` & `Content editor`
- Check the following menus: `VA Butler health care` & `VA Bay Pines health care`
- Enter the following paths (one per line): `/butler-health-care/pharmacy` & `*/work-with-us`
- Go to `/user/1730/roles` and check the following roles: `Content creator - VAMC` , `Content editor` , `Content reviewer` , `Content publisher`
- Go to `user/1730/workbench_access` and check `VA Altoona health care` & `VA Butler health care`
- Sign out, and login as vanessa.luxen@va.gov
- Go to `/node/2551/edit` click save, and visually verify that you cannot save, and see `The path alias you have created has already been reserved.` message
- Go to `/node/2560/edit` click save, and visually verify that you cannot save, and see `The path alias you have created has already been reserved.` message
- Go to `/node/1332/edit` and visually verify that you see an access denied message
- Go to `node/add/health_care_region_detail_page` open up the menu settings, and visually verify that only VA Butler vamc items are available, and all other vamc items are greyed out and unclickable
- Go to `/admin/structure/menu/` and visually verify access is denied.
- Go to `/admin/content?title=&type=All&moderation_state=All&owner=185` and visually verify you can only edit Detail pages.

### Definition of Done

- [ ] Documentation has been updated, if applicable.
- [ ] Tests have been added if necessary.
- [ ] Automated tests have passed.
- [ ] Code Quality Tests have passed.
- [ ] Acceptance Criteria in related issue are met.
- [ ] Manual Code Review Approved.

### Select Team for PR review

- [ ] `Core Application Team`
- [x] `Product Support Team`

### Is this PR blocked by another PR?

- [ ] `DO NOT MERGE`

### Does this PR need review from a Product Owner

- [ ] `Needs PO review`

### CMS user-facing annoucement

Is an announcement needed to let editors know of this change? 
- [ ] Yes, and it's written in issue ____ and queued for publication. 
  - [ ] Merge and ping @ rachel-kauff so she's ready to publish after deployment
- [ ] Yes, but it hasn't yet been written 
  - [ ] Don't merge yet -- ping @ rachel-kauff to prompt her to write and queue content
- [ ] No announcement is needed for this code change. 
  - [ ] Merge & carry on unburdened by announcements

### CODE REVIEW

**Comment by kevwalsh** [comment]
I had a look through. Looks like a lot of the ideas are there, but i think we need a sync on some of the ACs and overall purpose. 

- an editor at Butler should still be able to edit and save a node like /node/2551 with a "protected" path, but should not be able to move the menu item
- an editor at Butler should still be able to edit other Butler stuff
- content admins (or just admins, for this POC) need a way to specify which parents are "eligible" to have child menu items created via a detail page form.

i didn't read the code, but how is the list of menus at `va_gov_menu_access` populated? it only shows Health care menus (which is great). but wondering how that list is governed. 

i'm a bit concerned about the overlap between this and workbench_access_menu, which has been working properly for us (if not providing all the necessary features). 

We can discuss more later, maybe a 16th minute, but possibly a longer convo.

File: docroot/modules/custom/va_gov_menu_access/src/Routing/AdminRouteSubscriber.php [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+      'entity.menu.collection',
+      'entity.menu.edit_form',
+      'entity.menu.add_form',
+    ];
+    foreach ($collection->all() as $routename => $route) {
+      if (in_array($routename, $menu_routes)) {
+        $route->setRequirement(
+          '_custom_access',
+          '\Drupal\va_gov_menu_access\AccessChecks\RouteAccessChecks::access'
+          );
```
Comment (acrollet) [inline docroot/modules/custom/va_gov_menu_access/src/Routing/AdminRouteSubscriber.php:27]:
```suggestion
        );
```

File: docroot/modules/custom/va_gov_menu_access/va_gov_menu_access.module [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+use Drupal\Core\Access\AccessResult;
+use Drupal\Core\Form\FormStateInterface;
+use Drupal\Core\Session\AccountInterface;
+use Drupal\node\NodeInterface;
+
+/**
+ * Implements hook_form_alter().
+ */
+function va_gov_menu_access_form_alter(&$form, FormStateInterface $form_state, $form_id) {
+  $form['#attached']['library'][] = 'va_gov_menu_access/parent_section_select';
```
Comment (acrollet) [inline docroot/modules/custom/va_gov_menu_access/va_gov_menu_access.module:17]: Just checking if this JS should be added to all forms site-wide?

File: docroot/modules/custom/va_gov_menu_access/va_gov_menu_access.module [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    if ($admin_role_count < 1) {
+      $section_scheme = \Drupal::entityTypeManager()->getStorage('access_scheme')->load('section');
+      $user_sections = \Drupal::service('workbench_access.user_section_storage')->getUserSections($section_scheme, $account);
+      $node_section = $node->get('field_administration')->getString();
+      // User doesn't have section perms, so don't let them operate on node.
+      if (!in_array($node_section, $user_sections) && $op !== 'view') {
+        return AccessResult::forbidden();
+      }
+      // Not a detail page, and user is trying to operate on node.
+      if ($node->bundle() !== 'health_care_region_detail_page' && ($op === 'update' || $op === 'create')) {
```
Comment (acrollet) [inline docroot/modules/custom/va_gov_menu_access/va_gov_menu_access.module:76]: I think a comment about why `health_care_region_detail_page` is excluded would be helpful for our future selves...

**Review by acrollet — COMMENTED** [review]
Just a little feedback - I have one high-level question, which is if you've done any benchmarking with and without this module enabled? The school of hard knocks has taught me to be careful with anything that enforces access restrictions in Drupal... ;-)

File: docroot/modules/custom/va_gov_menu_access/js/va_gov_menu_access.es6.js [hunk 4]
```diff
@@ -0,0 +1,48 @@
+/**
+ * @file
+ */
+
+((Drupal, drupalSettings) => {
```
Comment (indytechcook) [inline docroot/modules/custom/va_gov_menu_access/js/va_gov_menu_access.es6.js:5]: Anything security-related should be on the server/php side.  Client-side Javascript is easy to change and hack around.  Plus this removes any Drupal form security.

File: docroot/modules/custom/va_gov_menu_access/va_gov_menu_access.module (hunk 2 above)
Comment (ethanteague) [reply docroot/modules/custom/va_gov_menu_access/va_gov_menu_access.module:17]: Nah, you are right to flag this - I'll lock this down, @acrollet

File: docroot/modules/custom/va_gov_menu_access/src/Plugin/Validation/Constraint/ReservedPathValidator.php [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+/**
+ * Validates the ReservedPath constraint.
+ */
+class ReservedPathValidator extends ConstraintValidator {
+
+  /**
+   * {@inheritdoc}
+   */
+  public function validate($item, Constraint $constraint) {
+    $current_user_roles = \Drupal::currentUser()->getRoles();
```
[the remark this answers is not in the document]
Comment (ethanteague) [reply docroot/modules/custom/va_gov_menu_access/src/Plugin/Validation/Constraint/ReservedPathValidator.php:17]: Yep - my bad, @acrollet - missed this.

File: tests/behat/features/google_tag_manager.feature [hunk 6]
```diff
@@ -80,9 +80,3 @@ Feature: Google Tag Manager dataLayer values are correct
     And my workbench access sections are set to "165,176,177,212,246,374,375,376,377,378"
     And I am on "/node/2"
     Then the GTM data layer value for "userSections" should be set to "165,176,177,212,246,374,375,376,377,378"
-    And my workbench access sections are set to "1,10,157,162,163,191,192,2,204,205,206,207,208,209,210,211,247,248,249,250,251,252,253,254,255,295,3,46,5,6,65,66,67,68,69,7,70,72,73,74,75,76,77,8"
-    And I am on "/node/2"
-    Then the GTM data layer value for "userSections" should be set to "1,10,157,162,163,191,192,2,204,205,206,207,208,209,210,211,247,248,249,250,251,252,253,254,255,295,3,46,5,6,65,66,67,68,69,7,70,72,73,74,75,76,77,8"
```
[the remark this answers is not in the document]
Comment (ethanteague) [reply tests/behat/features/google_tag_manager.feature:85]: Yeah, I was thinking that these were no longer going to be applicable with the direction this particular spike was moving in, @acrollet , but we just had another meeting about this work.
TLDR: this is still applicable, and I'm gonna put it back.

File: docroot/modules/custom/va_gov_menu_access/js/va_gov_menu_access.es6.js (hunk 4 above)
Comment (ethanteague) [reply docroot/modules/custom/va_gov_menu_access/js/va_gov_menu_access.es6.js:5]: We just chatted about this one - little bit of confusion about the purpose of this code. Will leave in for ux purposes, and add form validation.

File: docroot/modules/custom/va_gov_menu_access/va_gov_menu_access.module [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+}
+
+/**
+ * Implements hook_page_attachments().
+ *
+ * Adds menu access info to drupalsettings for js menu association limiting.
+ */
+function va_gov_menu_access_page_attachments(array &$attachments) {
+  $menu_access_settings = _va_gov_menu_access_get_menu_access();
+  $attachments['#attached']['drupalSettings']['vagovmenuaccess'] = [
```
[the remark this answers is not in the document]
Comment (ethanteague) [reply docroot/modules/custom/va_gov_menu_access/va_gov_menu_access.module:43]: > Just a little feedback - I have one high-level question, which is if you've done any benchmarking with and without this module enabled? The school of hard knocks has taught me to be careful with anything that enforces access restrictions in Drupal... ;-)

Great q! I haven't done any local profiling on this project - what tool do you think I could integrate kinda low-efforty into lando, @acrollet  ?

[the remark this answers is not in the document]
Comment (acrollet) [reply docroot/modules/custom/va_gov_menu_access/va_gov_menu_access.module:43]: sorry for the slow response! This week has been pretty crazy for me. I think blackfire would probably be a good starting point since we have the config already there: https://github.com/department-of-veterans-affairs/va.gov-cms/blob/e9e7cfd10a467a443bba115b68a1b68bf9190518/.lando.yml#L196 I think @ElijahLynn is the person to ask about getting a seat.

[the remark this answers is not in the document]
Comment (ElijahLynn) [reply docroot/modules/custom/va_gov_menu_access/va_gov_menu_access.module:43]: Ethan has a seat already, we need a team-wide training on how to use it. Ethan, if you need me to resend the invite lemme know.

**Comment by ethanteague** [comment]
Closing in favor of https://github.com/department-of-veterans-affairs/va.gov-cms/pull/4894
